File: desktop/native/SurveyController.App/Services/TaskNotification.cpp
#include "pch.h"
#include "TaskNotification.h"

namespace winrt::SurveyController::App::Services
{
    namespace
    {
        hstring EscapeXml(hstring const& value)
        {
            std::wstring escaped;
            escaped.reserve(value.size());
            for (auto const character : value)
            {
                switch (character)
                {
                case L'&': escaped += L"&amp;"; break;
                case L'<': escaped += L"&lt;"; break;
                case L'>': escaped += L"&gt;"; break;
                case L'\"': escaped += L"&quot;"; break;
                case L'\'': escaped += L"&apos;"; break;
                default: escaped += character; break;
                }
            }
            return hstring{ escaped };
        }
    }

    TaskNotification& TaskNotification::Current()
    {
        static TaskNotification instance;
        return instance;
    }

    bool TaskNotification::Show(hstring const& title, hstring const& body) noexcept
    {
        try
        {
            if (!EnsureRegistered()) return false;
            auto payload = hstring{
                L"<toast><visual><binding template=\"ToastGeneric\"><text>" +
                std::wstring{ EscapeXml(title) } + L"</text><text>" +
                std::wstring{ EscapeXml(body) } + L"</text></binding></visual></toast>"
            };
            auto notification = Microsoft::Windows::AppNotifications::AppNotification{ payload };
            Microsoft::Windows::AppNotifications::AppNotificationManager::Default().Show(notification);
            return true;
        }
        catch (...)
        {
            return false;
        }
    }

    bool TaskNotification::EnsureRegistered()
    {
        if (m_registered) return true;
        auto manager = Microsoft::Windows::AppNotifications::AppNotificationManager::Default();
        m_invokedToken = manager.NotificationInvoked([](auto const&, auto const&) {});
        manager.Register();
        m_registered = true;
        return true;
    }

    TaskNotification::~TaskNotification()
    {
        if (!m_registered) return;
        try
        {
            auto manager = Microsoft::Windows::AppNotifications::AppNotificationManager::Default();
            manager.NotificationInvoked(m_invokedToken);
            manager.Unregister();
        }
        catch (...)
        {
        }
    }
}

File: desktop/native-cs/src/SurveyController.App/Services/TaskNotification.cs
using System.Text;
using Microsoft.Windows.AppNotifications;

namespace SurveyController.App.Services;

/// <summary>任务完成/失败的系统 Toast 通知，语义与 C++ Services/TaskNotification 一致。</summary>
internal static class TaskNotification
{
    private static bool _registered;

    public static bool Show(string title, string body)
    {
        try
        {
            EnsureRegistered();
            var payload = new StringBuilder("<toast><visual><binding template=\"ToastGeneric\"><text>")
                .Append(EscapeXml(title))
                .Append("</text><text>")
                .Append(EscapeXml(body))
                .Append("</text></binding></visual></toast>")
                .ToString();
            AppNotificationManager.Default().Show(new AppNotification(payload));
            return true;
        }
        catch (Exception)
        {
            return false;
        }
    }

    private static void EnsureRegistered()
    {
        if (_registered)
        {
            return;
        }
        var manager = AppNotificationManager.Default();
        manager.NotificationInvoked += (_, _) =>
        {
            // 激活行为暂不处理；注册事件以维持与 C++ 相同的生命周期。
        };
        manager.Register();
        _registered = true;
    }

    private static string EscapeXml(string value)
    {
        var builder = new StringBuilder(value.Length);
        foreach (var character in value)
        {
            switch (character)
            {
                case '&': builder.Append("&amp;"); break;
                case '<': builder.Append("&lt;"); break;
                case '>': builder.Append("&gt;"); break;
                case '"': builder.Append("&quot;"); break;
                case '\'': builder.Append("&apos;"); break;
                default: builder.Append(character); break;
            }
        }
        return builder.ToString();
    }
}
